src/Pull.cpp

```diff
@@ -0,0 +1,200 @@
+#include "Pull.h"
+#include <stdexcept>
+#include <g3log/g3log.hpp>
+/**
+ * Construct a blocking nanomsg pull (no timeout)
+ */
+Pull::Pull(const std::string& location, const bool shouldConnect) {
+   mProtocolHandler = std::move(std::unique_ptr<NanoProtocol>(
+                                   new NanoProtocol(location)));
+   mSocket = nn_socket(AF_SP, NN_PULL);
+   if (mSocket < 0) {
+      throw std::runtime_error("could not open transport socket" +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect){
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_connect";
+   } else{
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_bind";
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   LOG(INFO) << connectionType
+              << " socket at: " << mProtocolHandler->GetLocation();
+}
+/**
+ * Construct a non-blocking nanomsg pull (timeout)
+ */
+Pull::Pull(const std::string& location,
+                 const int timeoutInMs,
+                 const bool shouldConnect) {
+   mProtocolHandler = std::move(std::unique_ptr<NanoProtocol>(
+                                   new NanoProtocol(location)));
+   mSocket = nn_socket(AF_SP, NN_PULL);
+   if (mSocket < 0) {
+      throw std::runtime_error("could not open transport socket" +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   nn_setsockopt(mSocket,
```

Comment:
My point here is that the socket will have some default options, right?
1) Do you know what these options are
2) Does it make sense to use the nn_setsockopt even for the default settings just to make it explicit what the behaviour is?
3) if 2) then it'll be better if you used ```delegating constructors```` which would reduce the duplicated code